File: stress-test.py
#!/usr/bin/env python3
"""KataGo 压力测试 - 模拟20个用户同时快速请求KataGo"""

import asyncio
import aiohttp
import time
import json
import sys

BASE_URL = sys.argv[1] if len(sys.argv) > 1 else "https://letusgoa.cn"
CONCURRENT_USERS = 20
MOVES_PER_USER = 3  # 每个用户模拟3步

# 预设落子序列（9路棋盘，避免不合法落子）
MOVE_SEQUENCES = [
    [{"row":2,"col":2,"color":"black"}],
    [{"row":2,"col":2,"color":"black"},{"row":6,"col":6,"color":"white"}],
    [{"row":2,"col":6,"color":"black"}],
    [{"row":6,"col":2,"color":"black"},{"row":2,"col":2,"color":"white"}],
    [{"row":4,"col":4,"color":"black"}],
    [{"row":2,"col":4,"color":"black"},{"row":6,"col":4,"color":"white"}],
    [{"row":4,"col":2,"color":"black"}],
    [{"row":4,"col":6,"color":"black"},{"row":2,"col":6,"color":"white"}],
    [{"row":0,"col":0,"color":"black"}],
    [{"row":8,"col":8,"color":"black"},{"row":0,"col":8,"color":"white"}],
    [{"row":0,"col":4,"color":"black"}],
    [{"row":8,"col":4,"color":"black"},{"row":4,"col":0,"color":"white"}],
    [{"row":8,"col":0,"color":"black"}],
    [{"row":0,"col":2,"color":"black"},{"row":8,"col":6,"color":"white"}],
    [{"row":6,"col":0,"color":"black"}],
    [{"row":2,"col":8,"color":"black"},{"row":6,"col":8,"color":"white"}],
    [{"row":3,"col":3,"color":"black"}],
    [{"row":5,"col":5,"color":"black"},{"row":3,"col":5,"color":"white"}],
    [{"row":5,"col":3,"color":"black"}],
    [{"row":1,"col":1,"color":"black"},{"row":7,"col":7,"color":"white"}],
]

results = []

async def simulate_user(session, user_id, token=None):
    """模拟单个用户的对弈请求"""
    moves = MOVE_SEQUENCES[user_id % len(MOVE_SEQUENCES)]
    headers = {"Content-Type": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    
    user_results = []
    for move_idx in range(MOVES_PER_USER):
        current_moves = moves * (move_idx + 1)  # 逐步增加落子
        payload = {
            "boardSize": 9,
            "difficulty": "hard",
            "engine": "katago",
            "moves": current_moves
        }
        
        start = time.time()
        try:
            async with session.post(
                f"{BASE_URL}/api/go-engine",
                json=payload,
                headers=headers,
                timeout=aiohttp.ClientTimeout(total=60)
            ) as resp:
                elapsed = time.time() - start
                status = resp.status
                data = await resp.json()
                engine_used = data.get("engine", "unknown")
                no_engine = data.get("noEngine", False)
                error = data.get("error", "")
                
        except asyncio.TimeoutError:
            elapsed = time.time() - start
            status = 0
            engine_used = "timeout"
            no_engine = False
            error = "timeout"
        except Exception as e:
            elapsed = time.time() - start
            status = -1
            engine_used = "error"
            no_engine = False
            error = str(e)[:50]
        
        result = {
            "user": user_id,
            "move": move_idx + 1,
            "status": status,
            "elapsed": round(elapsed, 2),
            "engine": engine_used,
            "noEngine": no_engine,
            "error": error
        }
        user_results.append(result)
        results.append(result)
        
        status_icon = "✅" if status == 200 else "❌"
        engine_info = engine_used if not no_engine else "local(fallback)"
        print(f"  用户{user_id:02d} 第{move_idx+1}步: {status_icon} {status} {elapsed:.1f}s [{engine_info}]")
        
        # 快速落子：不等太久就下下一步
        await asyncio.sleep(0.5)
    
    return user_results

async def check_queue(session):
    """持续监控排队状态"""
    for i in range(30):
        try:
            async with session.get(f"{BASE_URL}/api/go-engine", timeout=aiohttp.ClientTimeout(total=5)) as resp:
                data = await resp.json()
                q = data.get("queueLength", "?")
                p = data.get("isProcessing", "?")
                print(f"  📊 排队监控 #{i+1}: 排队人数={q} 处理中={p}")
        except:
            print(f"  📊 排队监控 #{i+1}: 获取失败")
        await asyncio.sleep(2)

async def main():
    print(f"=== KataGo 压力测试 ===")
    print(f"目标: {BASE_URL}")
    print(f"并发用户: {CONCURRENT_USERS}")
    print(f"每用户步数: {MOVES_PER_USER}")
    print()
    
    async with aiohttp.ClientSession() as session:
        # 注册测试用户获取token
        tokens = []
        for i in range(CONCURRENT_USERS):
            nickname = f"stresstest_{i:02d}"
            password = f"stress{i:02d}pass"
            try:
                async with session.post(
                    f"{BASE_URL}/api/auth/register",
                    json={"nickname": nickname, "password": password},
                    timeout=aiohttp.ClientTimeout(total=10)
                ) as resp:
                    data = await resp.json()
                    token = data.get("token", "")
                    if not token:
                        # 可能已注册，尝试登录
                        async with session.post(
                            f"{BASE_URL}/api/auth/login",
                            json={"nickname": nickname, "password": password},
                            timeout=aiohttp.ClientTimeout(total=10)
                        ) as resp2:
                            data2 = await resp2.json()
                            token = data2.get("token", "")
                    tokens.append(token)
            except Exception as e:
                tokens.append("")
                print(f"  注册用户{i:02d}失败: {e}")
        
        print(f"\n✅ 已准备 {sum(1 for t in tokens if t)} 个用户token\n")
        
        # 启动排队监控（后台）
        queue_task = asyncio.create_task(check_queue(session))
        
        # 同时启动所有用户
        print(f"🚀 开始压力测试 - {CONCURRENT_USERS}个用户同时请求KataGo...\n")
        start_time = time.time()
        
        tasks = [
            simulate_user(session, i, tokens[i] if i < len(tokens) else None)
            for i in range(CONCURRENT_USERS)
        ]
        await asyncio.gather(*tasks)
        
        total_time = time.time() - start_time
        queue_task.cancel()
        
        # 统计结果
        print(f"\n=== 压力测试结果 ===")
        print(f"总耗时: {total_time:.1f}s")
        print(f"总请求数: {len(results)}")
        
        success = [r for r in results if r["status"] == 200]
        failed = [r for r in results if r["status"] != 200]
        katago_used = [r for r in success if r["engine"] == "katago"]
        fallback = [r for r in success if r["noEngine"]]
        
        print(f"成功: {len(success)}  失败: {len(failed)}")
        print(f"KataGo处理: {len(katago_used)}  回退本地AI: {len(fallback)}")
        
        if success:
            times = [r["elapsed"] for r in success]
            print(f"响应时间: 最快{min(times):.1f}s  最慢{max(times):.1f}s  平均{sum(times)/len(times):.1f}s")
        
        if failed:
            print(f"\n失败请求:")
            for r in failed[:10]:
                print(f"  用户{r['user']:02d} 第{r['move']}步: status={r['status']} error={r['error']}")

if __name__ == "__main__":
    asyncio.run(main())

File: scripts/stress-test.js
/**
 * 双引擎压力测试脚本 v3
 *
 * 针对线上 letusgoa.cn，测试双引擎常驻架构的并发能力
 *
 * 三种测试模式：
 *   1. genmove  — 纯对弈引擎压力测试
 *   2. analyze  — 纯分析引擎压力测试（模拟大量用户同时点"教学提示"）
 *   3. mixed    — 混合压力（对弈 + 分析同时进行），验证两引擎互不阻塞
 *
 * 用法:
 *   node scripts/stress-test.js [模式] [BASE_URL]
 *
 * 示例:
 *   node scripts/stress-test.js genmove
 *   node scripts/stress-test.js analyze https://letusgoa.cn
 *   node scripts/stress-test.js mixed
 *
 * 环境变量:
 *   GENMOVE_USERS     对弈并发用户数，默认 10
 *   ANALYZE_USERS     分析并发用户数，默认 20
 *   BOARD_SIZE        棋盘大小，默认 19
 *   DIFFICULTY        easy | medium | hard，默认 hard
 *   STEP_INTERVAL_MS  每步间隔(ms)，默认 8000
 *   MAX_STEPS         每局最多步数，默认 15
 */

const MODE = process.argv[2] || 'mixed';
const BASE_URL = process.argv[3] || 'https://letusgoa.cn';

const GENMOVE_USERS = parseInt(process.env.GENMOVE_USERS || '10', 10);
const ANALYZE_USERS = parseInt(process.env.ANALYZE_USERS || '20', 10);
const BOARD_SIZE = parseInt(process.env.BOARD_SIZE || '19', 10);
const DIFFICULTY = process.env.DIFFICULTY || 'hard';
const STEP_INTERVAL_MS = parseInt(process.env.STEP_INTERVAL_MS || '8000', 10);
const MAX_STEPS = parseInt(process.env.MAX_STEPS || '15', 10);
const STAGGER_DELAY_MS = 300;
const REQUEST_TIMEOUT_MS = 120000;

// ---- 围棋最小逻辑 ----
function createEmptyBoard(size) {
  return Array.from({ length: size }, () => Array(size).fill(0));
}

function getNeighbors(row, col, size) {
  const neighbors = [];
  if (row > 0) neighbors.push([row - 1, col]);
  if (row < size - 1) neighbors.push([row + 1, col]);
  if (col > 0) neighbors.push([row, col - 1]);
  if (col < size - 1) neighbors.push([row, col + 1]);
  return neighbors;
}

function getGroup(board, row, col, size) {
  const color = board[row][col];
  if (color === 0) return { stones: [], liberties: 0 };
  const visited = new Set();
  const stones = [];
  const liberties = new Set();
  const stack = [[row, col]];
  while (stack.length > 0) {
    const [r, c] = stack.pop();
    const key = `${r},${c}`;
    if (visited.has(key)) continue;
    visited.add(key);
    stones.push([r, c]);
    for (const [nr, nc] of getNeighbors(r, c, size)) {
      const nKey = `${nr},${nc}`;
      if (board[nr][nc] === 0) {
        liberties.add(nKey);
      } else if (board[nr][nc] === color && !visited.has(nKey)) {
        stack.push([nr, nc]);
      }
    }
  }
  return { stones, liberties: liberties.size };
}

function removeGroup(board, stones) {
  for (const [r, c] of stones) board[r][c] = 0;
}

function isValidMove(board, row, col, color, size) {
  if (board[row][col] !== 0) return false;
  const testBoard = board.map(r => [...r]);
  testBoard[row][col] = color;
  const opponent = color === 1 ? 2 : 1;
  for (const [nr, nc] of getNeighbors(row, col, size)) {
    if (testBoard[nr][nc] === opponent) {
      const group = getGroup(testBoard, nr, nc, size);
      if (group.liberties === 0) removeGroup(testBoard, group.stones);
    }
  }
  const selfGroup = getGroup(testBoard, row, col, size);
  return selfGroup.liberties > 0;
}

function findRandomValidMove(board, color, size) {
  const validMoves = [];
  for (let r = 0; r < size; r++) {
    for (let c = 0; c < size; c++) {
      if (isValidMove(board, r, c, color, size)) {
        validMoves.push([r, c]);
      }
    }
  }
  if (validMoves.length === 0) return null;
  return validMoves[Math.floor(Math.random() * validMoves.length)];
}

function applyMove(board, row, col, color, size) {
  const newBoard = board.map(r => [...r]);
  newBoard[row][col] = color;
  const opponent = color === 1 ? 2 : 1;
  for (const [nr, nc] of getNeighbors(row, col, size)) {
    if (newBoard[nr][nc] === opponent) {
      const group = getGroup(newBoard, nr, nc, size);
      if (group.liberties === 0) removeGroup(newBoard, group.stones);
    }
  }
  return newBoard;
}

// ---- API 工具 ----
async function api(method, path, body, token) {
  const headers = { 'Content-Type': 'application/json' };
  if (token) headers['Authorization'] = `Bearer ${token}`;
  const opts = { method, headers, signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS) };
  if (body) opts.body = JSON.stringify(body);
  const res = await fetch(`${BASE_URL}${path}`, opts);
  return res.json();
}

// ---- 全局统计 ----
const stats = {
  genmove: { requests: 0, success: 0, errors: 0, passes: 0, totalTime: 0, maxTime: 0, delays: 0 },
  analyze: { requests: 0, success: 0, errors: 0, totalTime: 0, maxTime: 0, delays: 0, totalVisits: 0 },
  startTime: Date.now(),
};

let activeGenmove = 0;
let activeAnalyze = 0;
let completedGenmove = 0;
let completedAnalyze = 0;

function printProgress() {
  const elapsed = ((Date.now() - stats.startTime) / 1000).toFixed(0);
  const gm = stats.genmove;
  const an = stats.analyze;
  const gmAvg = gm.requests > 0 ? (gm.totalTime / gm.requests / 1000).toFixed(1) : '-';
  const anAvg = an.requests > 0 ? (an.totalTime / an.requests / 1000).toFixed(1) : '-';
  console.log(
    `[${elapsed}s] 活跃: GM=${activeGenmove} AN=${activeAnalyze} | ` +
    `GM: ${gm.requests}请求 ${gm.success}成功 ${gmAvg}s均耗 ${gm.delays}延迟 | ` +
    `AN: ${an.requests}请求 ${an.success}成功 ${anAvg}s均耗 ${an.delays}延迟 | ` +
    `完成: GM=${completedGenmove}/${MODE === 'analyze' ? 0 : GENMOVE_USERS} ` +
    `AN=${completedAnalyze}/${MODE === 'genmove' ? 0 : ANALYZE_USERS}`
  );
}

// ---- 对弈压力测试（genmove 引擎） ----
async function runGenmoveUser(userId) {
  const tag = `[GM${userId}]`;
  const userIdx = userId;
  let token;
  let gameStarted = false;

  try {
    const rid = Math.random().toString(36).slice(2, 5);
    const nickname = `gm${userIdx}_${rid}`;
    let regResult = await api('POST', '/api/auth/register', { nickname, password: 'test1234' });
    if (!regResult.token) {
      regResult = await api('POST', '/api/auth/register', { nickname: nickname + 'x', password: 'test1234' });
    }
    if (!regResult.token) throw new Error(`注册失败: ${regResult.error}`);
    token = regResult.token;

    gameStarted = true;
    activeGenmove++;
    console.log(`${tag} 开始对弈`);

    const gameResult = await api('POST', '/api/games', {
      boardSize: BOARD_SIZE, difficulty: DIFFICULTY, engine: 'katago',
      title: `压测-GM-${BOARD_SIZE}路-${DIFFICULTY}`, moves: [], status: 'playing',
    }, token);
    const gameId = gameResult.game?.id || gameResult.id;

    let board = createEmptyBoard(BOARD_SIZE);
    let moves = [];
    let playerColor = 1;

    for (let step = 1; step <= MAX_STEPS; step++) {
      const move = findRandomValidMove(board, playerColor, BOARD_SIZE);
      if (!move) break;
      const [row, col] = move;
      moves.push({ row, col, color: playerColor });
      board = applyMove(board, row, col, playerColor, BOARD_SIZE);

      const t0 = Date.now();
      let engineResult;
      try {
        engineResult = await api('POST', '/api/go-engine', {
          boardSize: BOARD_SIZE, difficulty: DIFFICULTY, engine: 'katago', moves,
        }, token);
      } catch (e) {
        stats.genmove.errors++;
        console.log(`${tag} 第${step}步 异常: ${e.message}`);
        break;
      }
      const elapsed = Date.now() - t0;

      stats.genmove.requests++;
      stats.genmove.totalTime += elapsed;
      if (elapsed > stats.genmove.maxTime) stats.genmove.maxTime = elapsed;
      if (elapsed > STEP_INTERVAL_MS) stats.genmove.delays++;

      const delayMark = elapsed > STEP_INTERVAL_MS ? ' ⏳' : '';
      console.log(`${tag} #${step} ${elapsed}ms${delayMark}`);

      if (engineResult.error && !engineResult.insufficientPoints) {
        stats.genmove.errors++;
        break;
      }
      if (engineResult.insufficientPoints) {
        console.log(`${tag} 积分不足`);
        break;
      }

      if (engineResult.move) {
        const aiColor = playerColor === 1 ? 2 : 1;
        board = applyMove(board, engineResult.move.row, engineResult.move.col, aiColor, BOARD_SIZE);
        moves.push({ row: engineResult.move.row, col: engineResult.move.col, color: aiColor });
        stats.genmove.success++;
      } else if (engineResult.pass) {
        stats.genmove.passes++;
        stats.genmove.success++;
        console.log(`${tag} AI停手`);
        break;
      }

      if (step % 5 === 0 && gameId) {
        await api('PUT', `/api/games/${gameId}`, { moves, status: 'playing' }, token).catch(() => {});
      }

      if (step < MAX_STEPS) {
        await new Promise(r => setTimeout(r, Math.max(0, STEP_INTERVAL_MS - elapsed)));
      }
    }

    if (gameId) {
      await api('PUT', `/api/games/${gameId}`, { moves, status: 'finished' }, token).catch(() => {});
    }
  } catch (e) {
    stats.genmove.errors++;
    console.log(`${tag} 异常: ${e.message}`);
  } finally {
    if (gameStarted) activeGenmove--;
    completedGenmove++;
  }
}

// ---- 分析压力测试（analyze 引擎） ----
async function runAnalyzeUser(userId) {
  const tag = `[AN${userId}]`;
  const userIdx = userId + 200;
  let token;
  let started = false;

  try {
    const rid = Math.random().toString(36).slice(2, 5);
    const nickname = `an${userIdx}_${rid}`;
    let regResult = await api('POST', '/api/auth/register', { nickname, password: 'test1234' });
    if (!regResult.token) {
      regResult = await api('POST', '/api/auth/register', { nickname: nickname + 'x', password: 'test1234' });
    }
    if (!regResult.token) throw new Error(`注册失败: ${regResult.error}`);
    token = regResult.token;

    started = true;
    activeAnalyze++;
    console.log(`${tag} 开始分析压测`);

    // 预构造一段棋谱
    let board = createEmptyBoard(BOARD_SIZE);
    let moves = [];
    let color = 1;
    for (let i = 0; i < 8; i++) {
      const move = findRandomValidMove(board, color, BOARD_SIZE);
      if (!move) break;
      const [row, col] = move;
      moves.push({ row, col, color });
      board = applyMove(board, row, col, color, BOARD_SIZE);
      color = color === 1 ? 2 : 1;
    }

    // 循环点击"教学提示"，模拟真实用户行为
    for (let round = 1; round <= MAX_STEPS; round++) {
      // 每轮追加一手，让局面变化
      if (round > 1) {
        const move = findRandomValidMove(board, color, BOARD_SIZE);
        if (move) {
          const [row, col] = move;
          moves.push({ row, col, color });
          board = applyMove(board, row, col, color, BOARD_SIZE);
          color = color === 1 ? 2 : 1;
        }
      }

      const t0 = Date.now();
      let result;
      try {
        result = await api('POST', '/api/go-engine', {
          action: 'analyze',
          boardSize: BOARD_SIZE,
          moves,
          difficulty: DIFFICULTY,
        }, token);
      } catch (e) {
        stats.analyze.errors++;
        console.log(`${tag} #${round} 异常: ${e.message}`);
        break;
      }
      const elapsed = Date.now() - t0;

      stats.analyze.requests++;
      stats.analyze.totalTime += elapsed;
      if (elapsed > stats.analyze.maxTime) stats.analyze.maxTime = elapsed;
      if (elapsed > STEP_INTERVAL_MS) stats.analyze.delays++;

      const visits = result?.analysis?.actualVisits || 0;
      stats.analyze.totalVisits += visits;

      const ok = !!result?.analysis;
      if (ok) stats.analyze.success++;
      else stats.analyze.errors++;

      const delayMark = elapsed > STEP_INTERVAL_MS ? ' ⏳' : '';
      console.log(`${tag} #${round} ${elapsed}ms visits=${visits} ${ok ? '✓' : '✗'}${delayMark}`);

      if (result?.insufficientPoints) {
        console.log(`${tag} 积分不足`);
        break;
      }

      await new Promise(r => setTimeout(r, Math.max(0, STEP_INTERVAL_MS - elapsed)));
    }
  } catch (e) {
    stats.analyze.errors++;
    console.log(`${tag} 异常: ${e.message}`);
  } finally {
    if (started) activeAnalyze--;
    completedAnalyze++;
  }
}

// ---- 主流程 ----
async function main() {
  console.log('='.repeat(70));
  console.log('  双引擎压力测试 v3');
  console.log('  目标:', BASE_URL);
  console.log('  模式:', MODE);
  console.log('  棋盘:', BOARD_SIZE, '路', DIFFICULTY);
  console.log('  对弈用户:', MODE === 'analyze' ? 0 : GENMOVE_USERS);
  console.log('  分析用户:', MODE === 'genmove' ? 0 : ANALYZE_USERS);
  console.log('  每步间隔:', STEP_INTERVAL_MS, 'ms | 最大步数:', MAX_STEPS);
  console.log('='.repeat(70));

  // 检查引擎状态
  try {
    const status = await api('GET', '/api/go-engine');
    const kt = status.engines?.find(e => e.id === 'katago');
    console.log(`  KataGo: ${kt?.available ? '✓ 可用' : '✗ 不可用'}`);
    console.log(`  对弈模型: ${status.katago?.gameModel?.name || '-'}`);
    console.log(`  分析模型: ${status.katago?.analysisModel?.name || '-'}`);
  } catch (e) {
    console.log(`  无法获取引擎状态: ${e.message}`);
  }
  console.log('');

  const progressTimer = setInterval(printProgress, 15000);
  const startTime = Date.now();

  const allPromises = [];
  let launchIndex = 0;

  if (MODE !== 'analyze') {
    for (let i = 1; i <= GENMOVE_USERS; i++) {
      launchIndex++;
      const delay = launchIndex * STAGGER_DELAY_MS;
      allPromises.push(
        new Promise(r => setTimeout(r, delay)).then(() => runGenmoveUser(i))
      );
    }
  }

  if (MODE !== 'genmove') {
    for (let i = 1; i <= ANALYZE_USERS; i++) {
      launchIndex++;
      const delay = launchIndex * STAGGER_DELAY_MS;
      allPromises.push(
        new Promise(r => setTimeout(r, delay)).then(() => runAnalyzeUser(i))
      );
    }
  }

  console.log(`已调度 ${launchIndex} 个用户，错峰启动中...`);
  console.log(`预计全部启动: ~${((launchIndex * STAGGER_DELAY_MS) / 1000).toFixed(1)}s`);
  console.log('');

  await Promise.allSettled(allPromises);
  clearInterval(progressTimer);

  const totalElapsed = (Date.now() - startTime) / 1000;

  printProgress();
  console.log('');
  console.log('='.repeat(70));
  console.log('  压力测试完成');
  console.log('='.repeat(70));
  console.log(`  总耗时: ${totalElapsed.toFixed(1)}s`);

  // 对弈统计
  if (stats.genmove.requests > 0) {
    const s = stats.genmove;
    const avg = (s.totalTime / s.requests / 1000).toFixed(2);
    const p99 = s.maxTime / 1000;
    const rps = (s.requests / totalElapsed).toFixed(1);
    console.log('');
    console.log('  --- 对弈引擎 (genmove) ---');
    console.log(`  请求: ${s.requests} | 成功: ${s.success} | 错误: ${s.errors} | AI停手: ${s.passes}`);
    console.log(`  平均: ${avg}s | 最大: ${p99.toFixed(1)}s | 延迟: ${s.delays}次 | RPS: ${rps}`);
    console.log(`  成功率: ${((s.success / s.requests) * 100).toFixed(1)}%`);
  }

  // 分析统计
  if (stats.analyze.requests > 0) {
    const s = stats.analyze;
    const avg = (s.totalTime / s.requests / 1000).toFixed(2);
    const p99 = s.maxTime / 1000;
    const avgVisits = Math.round(s.totalVisits / s.requests);
    const rps = (s.requests / totalElapsed).toFixed(1);
    console.log('');
    console.log('  --- 分析引擎 (analyze) ---');
    console.log(`  请求: ${s.requests} | 成功: ${s.success} | 错误: ${s.errors}`);
    console.log(`  平均: ${avg}s | 最大: ${p99.toFixed(1)}s | 延迟: ${s.delays}次 | RPS: ${rps}`);
    console.log(`  平均 visits: ${avgVisits}`);
    console.log(`  成功率: ${((s.success / s.requests) * 100).toFixed(1)}%`);
  }

  // 关键结论
  console.log('');
  console.log('  --- 关键结论 ---');
  if (MODE === 'mixed') {
    const gmBlocked = stats.genmove.delays > 0;
    const anBlocked = stats.analyze.delays > 0;
    if (!gmBlocked && !anBlocked) {
      console.log('  ✓ 两引擎互不阻塞（无互相延迟）');
    } else {
      console.log(`  ${gmBlocked ? '✗' : '✓'} 对弈引擎${gmBlocked ? '有' : '无'}延迟`);
      console.log(`  ${anBlocked ? '✗' : '✓'} 分析引擎${anBlocked ? '有' : '无'}延迟`);
    }
  }
  console.log('='.repeat(70));
}

main().catch(console.error);
